td-exception/src/idt.rs

```diff
@@ -0,0 +1,207 @@
+// Copyright (c) 2021 Intel Corporation
+//
+// SPDX-License-Identifier: BSD-2-Clause-Patent
+
+//! Manipulate x86_64 Interrupt Descriptor Table (IDT).
+//!
+//! Setup a stub IDT for td-shim, which assumes 1:1 mapping between physical address and virtual
+//! address in identity mapping mode.
+//!
+//! It also handles Virtualization Interrupt for Intel TDX technology.
+
+use core::mem::{self, size_of};
+use core::slice::from_raw_parts_mut;
+
+use bitflags::bitflags;
+use lazy_static::lazy_static;
+
+use crate::interrupt;
+
+const IDT_ENTRY_COUNT: usize = 256;
+
+extern "win64" {
+    fn sidt_call(addr: usize);
+    fn lidt_call(addr: usize);
+    fn read_cs_call() -> u16;
+}
+
+lazy_static! {
+    static ref INIT_IDT: Idt = Idt::new();
+}
+
+#[repr(C, packed)]
+pub struct DescriptorTablePointer {
+    limit: u16,
+    base: u64,
+}
+
+#[no_mangle]
+/// # Safety
+///
+/// This function is unsafe because of the lidt_call()
+pub unsafe fn init() {
+    let mut idtr = DescriptorTablePointer { limit: 1, base: 0 };
+    let current_idt = &INIT_IDT.entries;
+
+    idtr.limit = (current_idt.len() * mem::size_of::<IdtEntry>() - 1) as u16;
+    idtr.base = current_idt.as_ptr() as u64;
+
+    lidt_call(&idtr as *const DescriptorTablePointer as usize);
+}
+
+pub type IdtEntries = [IdtEntry; IDT_ENTRY_COUNT];
+
+// 8 alignment required
+#[repr(C, align(8))]
+pub struct Idt {
+    pub entries: IdtEntries,
+}
+
+impl Default for Idt {
+    fn default() -> Self {
+        Self::new()
+    }
+}
+
+impl Idt {
+    pub fn new() -> Self {
+        let mut idt = Self {
+            entries: [IdtEntry::new(); IDT_ENTRY_COUNT],
+        };
+        idt.init();
+        idt
+    }
+
+    pub fn init(&mut self) {
+        let current_idt = &mut self.entries;
+
+        // Set up exceptions handler according to Intel64 & IA32 Software Developer Manual
+        // Reference: https://www.intel.com/content/www/us/en/developer/articles/technical/intel-sdm.html
+        current_idt[0].set_func(interrupt::divide_by_zero);
+        current_idt[1].set_func(interrupt::debug);
+        current_idt[2].set_func(interrupt::non_maskable);
+        current_idt[3].set_func(interrupt::breakpoint);
+        current_idt[4].set_func(interrupt::overflow);
+        current_idt[5].set_func(interrupt::bound_range);
+        current_idt[6].set_func(interrupt::invalid_opcode);
+        current_idt[7].set_func(interrupt::device_not_available);
+        current_idt[8].set_func(interrupt::double_fault);
+        // 9 no longer available
+        current_idt[9].set_func(interrupt::default_exception);
+        current_idt[10].set_func(interrupt::invalid_tss);
+        current_idt[11].set_func(interrupt::segment_not_present);
+        current_idt[12].set_func(interrupt::stack_segment);
+        current_idt[13].set_func(interrupt::protection);
+        current_idt[14].set_func(interrupt::page);
+        // 15 reserved
+        current_idt[15].set_func(interrupt::default_exception);
```

Comment:
It would be great if the default handlers could be passed the index so that if they do get called, it's clear which idt resulted in the call. However, making that happens doesn't look very easy currently.

Reply:
That need a big change for the exception handler.
Please aware that the index is not pushed by CPU. It must be recorded by exception handler.

That means we need register different exception handler for different IDT index.
Is that sth you really want to do?

Reply:
It was just a thought, so feel free to ignore it :smile: I don't think it's required, but it would be a "nice to have".

However, fwics it would be hard to achieve since the macros create the handlers before `set_func()` gets called meaning we can't embed the index into the auto-generated handler without changing the way the code works currently.

Reply:
If we really want to do it, we can do it.

Actually, I did the work for the physical BIOS long times ago, with assumption that hardware may have bug and trigger some weird exception or interrupt sporadically at power on phase. What we do is: 
1) Write a common code to handle exception/interrupt.
2) Write a small portion of assembly code to push the index for every handler (duplicate 256 times, with fixed size entry), then jump to this common code.
3) Register the IDT for the 256 individual handler. Since the size of entry code is fixed, we just use a for-loop to register the (base address + index * sizeof(entry code)).
I am not sure if it is still a good idea for Rust. We register an address, not a function.

I also understand why it is good, because in case it happens, we can immediate know which exception/interrupt is triggered.
But I am not sure if it is worth to do in this td-shim project. We assume the usage of td-shim is simple, and we assume the VMM will not do crazy thing purposely.

I suggest we keep it as is at this moment.
If people has strong desire to add, or we do encounter real use case requiring use to dump index, then we add such support at a separate patch as enhancement.